File: ProjectTrackercs10/ProjectTracker.Library/ResourceAssignment.cs
using System;
using System.Data;
using System.Data.SqlClient;
using CSLA;
using CSLA.Data;

namespace ProjectTracker.Library
{
  [Serializable()]
  public class ResourceAssignment : Assignment
	{
    Guid _ProjectID = Guid.Empty;
    string _ProjectName = string.Empty;

    #region Business Properties and Methods

    public Guid ProjectID
    {
      get
      {
        return _ProjectID;
      }
    }

    public string ProjectName
    {
      get
      {
        return _ProjectName;
      }
    }

    public Project GetProject()
    {
      return Project.GetProject(_ProjectID);
    }

    #endregion

    #region System.Object Overrides

    public override string ToString()
    {
      return _ProjectName;
    }

    public new static bool Equals(object objA, object objB)
    {
      if(objA is ResourceAssignment && objB is ResourceAssignment)
        return ((ResourceAssignment)objA).Equals((ResourceAssignment)objB);
      else
        return false;
    }

    public override bool Equals(object resourceAssignment)
    {
      if(resourceAssignment is ResourceAssignment)
        return Equals((ResourceAssignment)resourceAssignment);
      else
        return false;
    }

    public bool Equals(ResourceAssignment assignment)
    {
      return _ProjectID.Equals(assignment.ProjectID);
    }

    public override int GetHashCode()
    {
      return _ProjectID.GetHashCode();
    }

    #endregion

    #region Static Methods

    internal static ResourceAssignment NewResourceAssignment(Project project, string role) 
    {
        return new ResourceAssignment(project, role);
    }

    internal static ResourceAssignment NewResourceAssignment(Guid projectID, string role)
    {
      return new ResourceAssignment(Project.GetProject(projectID), role);
    }

    internal static ResourceAssignment NewResourceAssignment(Guid projectID)
    {
      return new ResourceAssignment(Project.GetProject(projectID), DefaultRole);
    }

    internal static ResourceAssignment GetResourceAssignment(SafeDataReader dr)
    {
      ResourceAssignment child = new ResourceAssignment();
      child.Fetch(dr);
      return child;
    }

    #endregion

    #region Constructors 

    private ResourceAssignment(Project project, string role)
    {
      _ProjectID = project.ID;
      _ProjectName = project.Name;
      _Assigned.Date = DateTime.Now;
      _Role = Convert.ToInt32(Roles.Key(role));
      MarkAsChild();
    }

    private ResourceAssignment()
    {
      // prevent direct creation of this object
    }

    #endregion

    #region Data Access

    private void Fetch(SafeDataReader dr)
    {
      _ProjectID = dr.GetGuid(0);
      _ProjectName = dr.GetString(1);
      _Assigned.Date = dr.GetDateTime(2);
      _Role = dr.GetInt32(3);
      MarkOld();
    }

    internal void Update(SqlTransaction tr, Resource resource)
    {
      // if we're not dirty then don't update the database
      if(!this.IsDirty)
        return;

      // do the update 
      SqlCommand cm = new SqlCommand();
      cm.Connection = tr.Connection;
      cm.Transaction = tr;
      cm.CommandType = CommandType.StoredProcedure;

      if(this.IsDeleted)
      {
        if(!this.IsNew)
        {
          // we're not new, so delete
          cm.CommandText = "deleteAssignment";
          cm.Parameters.Add("@ProjectID", _ProjectID);
          cm.Parameters.Add("@ResourceID", resource.ID);

          cm.ExecuteNonQuery();

          MarkNew();
        }
      }
      else
      {
        // we are either adding or updating
        if(this.IsNew)
        {
          // we're new, so insert
          cm.CommandText = "addAssignment";
        }
        else
        {
          // we're not new, so update
          cm.CommandText = "updateAssignment";
        }
        cm.Parameters.Add("@ProjectID", _ProjectID);
        cm.Parameters.Add("@ResourceID", resource.ID);
        cm.Parameters.Add("@Assigned", _Assigned.DBValue);
        cm.Parameters.Add("@Role", _Role);
        cm.ExecuteNonQuery();

        MarkOld();
      }
    }

    #endregion

	}
}

File: ProjectTracker10/ProjectTracker.Library/ResourceAssignment.vb
Imports System.Data.SqlClient
Imports CSLA.Data

<Serializable()> _
Public Class ResourceAssignment
  Inherits Assignment

  Private mProjectID As Guid = Guid.Empty
  Private mProjectName As String = ""

#Region " Business Properties and Methods "

  Public ReadOnly Property ProjectID() As Guid
    Get
      Return mProjectID
    End Get
  End Property

  Public ReadOnly Property ProjectName() As String
    Get
      Return mProjectName
    End Get
  End Property

  Public Function GetProject() As Project
    Return Project.GetProject(mProjectID)
  End Function

#End Region

#Region " System.Object Overrides "

  Public Overrides Function ToString() As String
    Return mProjectName
  End Function

  Public Overloads Shared Function Equals(ByVal objA As Object, ByVal objB As Object) As Boolean
    If TypeOf objA Is ResourceAssignment AndAlso TypeOf objB Is ResourceAssignment Then
      Return DirectCast(objA, ResourceAssignment).Equals(DirectCast(objB, ResourceAssignment))

    Else
      Return False
    End If
  End Function

  Public Overloads Overrides Function Equals(ByVal ResourceAssignment As Object) As Boolean
    If TypeOf ResourceAssignment Is ResourceAssignment Then
      Return Equals(DirectCast(ResourceAssignment, ResourceAssignment))

    Else
      Return False
    End If
  End Function

  Public Overloads Function Equals(ByVal Assignment As ResourceAssignment) As Boolean
    Return mProjectID.Equals(Assignment.ProjectID)
  End Function

  Public Overrides Function GetHashCode() As Integer
    Return mProjectID.GetHashCode
  End Function

#End Region

#Region " Shared Methods "

  Friend Shared Function NewResourceAssignment(ByVal Project As Project, ByVal Role As String) As ResourceAssignment
    Return New ResourceAssignment(Project, Role)
  End Function

  Friend Shared Function NewResourceAssignment(ByVal ProjectID As Guid, ByVal Role As String) As ResourceAssignment
    Return New ResourceAssignment(Project.GetProject(ProjectID), Role)
  End Function

  Friend Shared Function NewResourceAssignment(ByVal ProjectID As Guid) As ResourceAssignment
    Return New ResourceAssignment(Project.GetProject(ProjectID), DefaultRole)
  End Function

  Friend Shared Function GetResourceAssignment(ByVal dr As SafeDataReader) As ResourceAssignment
    Dim child As New ResourceAssignment
    child.Fetch(dr)
    Return child
  End Function

  Private Sub New(ByVal Project As Project, ByVal Role As String)
    mProjectID = Project.ID
    mProjectName = Project.Name
    mAssigned.Date = Now
    mRole = CInt(Roles.Key(Role))
    MarkAsChild()
  End Sub

  Private Sub New()
    ' prevent direct creation of this object
    MarkAsChild()
  End Sub

#End Region

#Region " Data Access "

  Private Sub Fetch(ByVal dr As SafeDataReader)
    With dr
      mProjectID = .GetGuid(0)
      mProjectName = .GetString(1)
      mAssigned.Date = .GetDateTime(2)
      mRole = .GetInt32(3)
      MarkOld()
    End With
  End Sub

  Friend Sub Update(ByVal tr As SqlClient.SqlTransaction, ByVal Resource As Resource)
    ' if we're not dirty then don't update the database
    If Not Me.IsDirty Then Exit Sub

    ' do the update 
    Dim cm As New SqlClient.SqlCommand
    With cm
      .Connection = tr.Connection
      .Transaction = tr
      .CommandType = CommandType.StoredProcedure

      If Me.IsDeleted Then
        If Not Me.IsNew Then
          ' we're not new, so delete
          .CommandText = "deleteAssignment"
          .Parameters.Add("@ProjectID", mProjectID)
          .Parameters.Add("@ResourceID", Resource.ID)

          .ExecuteNonQuery()

          MarkNew()
        End If

      Else
        ' we are either adding or updating
        If Me.IsNew Then
          ' we're new, so insert
          .CommandText = "addAssignment"

        Else
          ' we're not new, so update
          .CommandText = "updateAssignment"

        End If
        .Parameters.Add("@ProjectID", mProjectID)
        .Parameters.Add("@ResourceID", Resource.ID)
        .Parameters.Add("@Assigned", mAssigned.DBValue)
        .Parameters.Add("@Role", mRole)

        .ExecuteNonQuery()

        MarkOld()

      End If
    End With

  End Sub

#End Region

End Class
